godot-macros/src/class/data_models/func.rs

```diff
@@ -252,22 +257,49 @@ fn make_forwarding_closure(
     }
 }
 
-pub(crate) fn get_signature_info(signature: &venial::Function, has_gd_self: bool) -> SignatureInfo {
+/// Maps each usage of `Self` to the struct it's referencing,
+/// since `Self` can't be used inside nested functions.
+fn map_self_to_class_name<T: FromIterator<TokenTree>>(
+    tokens: impl IntoIterator<Item = TokenTree>,
+    class_name: &Ident,
+) -> T {
+    tokens
+        .into_iter()
+        .map(|tt| match tt {
+            // Change instances of Self to the class name.
+            TokenTree::Ident(ident) if ident == "Self" => TokenTree::Ident(class_name.clone()),
+            // Recurse into groups and make sure ALL instances are changed.
+            TokenTree::Group(group) => TokenTree::Group(Group::new(
+                group.delimiter(),
+                map_self_to_class_name(group.stream(), class_name),
+            )),
+            // Pass all other tokens through unchanged.
+            tt => tt,
+        })
+        .collect()
+}
+
+pub(crate) fn into_signature_info(
+    class_name: &Ident,
+    signature: venial::Function,
+    has_gd_self: bool,
+) -> SignatureInfo {
```

Comment:
As mentioned before, I'd make `signature` the first parameter since it is now the "subject" of the "into" operation.